### PR TITLE
Fixed 'newton:inertia' body-frame transform for inertial 'origin.rpy'

### LINKED ISSUE #131: inertia transform for URDF
### Version

v0.3.0

### Did you install official release artifacts or build from source?

Official release artifacts

### Did you modify the sdist, whl, or source code (including the pyproject.toml or uv.lock)?

No

### Python Version

3.12.3

### Operating System

ubuntu 24.04

### Describe the bug.

Even though it's rare, sometimes URDF can specify a transform on the inertial frame:
```
    <inertial>
      <mass value="3.87"/>
      <origin rpy="0 1.57079632679 0" xyz="-0.285775 0.0 0.0393"/>
      <inertia ixx="0.110793025489" ixy="0.0" ixz="0.0" iyy="0.110793025489" iyz="0.0" izz="0.010884375"/>
    </inertial>
```
from: 
https://github.com/Daniella1/urdf_files_dataset/blob/main/urdf_files/matlab/ur_e_description/urdf/universalUR10e.urdf#L180

However, I think the urdf converter is taking in the `inertial` as `newton:inertia` directly? https://github.com/newton-physics/urdf-usd-converter/blob/main/urdf_usd_converter/_impl/link.py#L128

### Minimum reproducible example

```python

```

### Relevant log output

```shell

```

### Code of Conduct

- [x] I agree to follow urdf-usd-converter Code of Conduct
- [x] I have searched the [open issues](https://github.com/newton-physics/urdf-usd-converter/issues) and have found no duplicates

### PR DESCRIPTION
**Pull request by ft-lab**
Fixes #131 

## Description

- Fix authoring of `newton:inertia` so the URDF inertia tensor is transformed into the body/link local frame when `inertial/origin.rpy` is set (previously the URDF values were copied verbatim).
- URDF `<inertia>` is expressed in the inertial frame defined by `<origin>`; `newton:inertia` must be in the body's local frame, so apply `I_body = R * I_urdf * R^T` using `origin.rpy`.
- Updated concept mapping docs.

## Unit test

- tests/testPhysicsInertia.py - TestPhysicsInertia - test_physics_inertia

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/newton-physics/urdf-usd-converter/blob/HEAD/CONTRIBUTING.md).

### CODE REVIEW

**Review by andrewkaufman — COMMENTED** [review]
Reviewed the math — the core transform is correct. Detailed feedback follows as inline comments.

File: urdf_usd_converter/_impl/link.py [hunk 1]
```diff
@@ -114,6 +115,30 @@ def convert_link(parent: Usd.Prim, having_articulation_root: bool, link: Element
...
 
 
+def _inertia_tensor_in_body_frame(inertia: ElementInertia, origin: ElementPose | None) -> list[float]:
+    """
+    Transform the URDF inertia tensor into the body's local frame.
+
+    URDF `inertia` is expressed in the inertial frame defined by `origin`.
+    `newton:inertia` requires the symmetric tensor in the body/link frame, so
+    when `origin.rpy` is non-identity the tensor is rotated as
+    `I_body = R^T * I_urdf * R` using `Gf`'s row-vector matrix convention
```
Comment (andrewkaufman) [inline urdf_usd_converter/_impl/link.py]: This docstring describes the literal `Gf` expression, but in standard notation `I_body = R^T * I_urdf * R` is the *inverse* rotation — the opposite of what the code correctly does. The PR description has it right (`I_body = R * I_urdf * R^T`).

Worth reconciling so the next reader doesn't "fix" this in the wrong direction. Something like:

> In `Gf`'s row-vector convention `Gf.Matrix3d(quat)` is the transpose of the column-convention `R`, so `R_gf.GetTranspose() * I * R_gf` evaluates to the standard `R * I * R^T`.

File: urdf_usd_converter/_impl/link.py [hunk 2]
```diff
@@ -114,6 +115,30 @@ def convert_link(parent: Usd.Prim, having_articulation_root: bool, link: Element
...
+    ixx = inertia.get_with_default("ixx")
+    ixy = inertia.get_with_default("ixy")
+    ixz = inertia.get_with_default("ixz")
+    iyy = inertia.get_with_default("iyy")
+    iyz = inertia.get_with_default("iyz")
+    izz = inertia.get_with_default("izz")
+    # Gf.Matrix3d is row-major: (r0c0, r0c1, r0c2, r1c0, ...)
+    mat = Gf.Matrix3d(ixx, ixy, ixz, ixy, iyy, iyz, ixz, iyz, izz)
+    if origin:
+        rotation = Gf.Matrix3d(float3_to_quatf(origin.get_with_default("rpy")))
```
Comment (andrewkaufman) [inline urdf_usd_converter/_impl/link.py]: `float3_to_quatf` returns `Gf.Quatf` (float32), so this rotation matrix is float32 even though `mat` is a `Gf.Matrix3d` and the attribute is `double[]`. The authored values land ~7 significant digits from correct:

```
exact (float64) Ixx : 1.1744663602719603
authored (Quatf)    : 1.1744663749117963
max abs error       : 4.269e-08
trace drift         : +8.412e-08
```

The trace of an inertia tensor is invariant under a similarity transform, so that drift is pure numerical error rather than a property of the rotation.

This matters because `newton:inertia`'s schema doc explicitly says *"Double precision prevents lossy conversion from source data that already represents the full tensor"*, and the 0.2.0 changelog introduced the attribute as *"allowing exact URDF inertia to be round-tripped without decomposition."* As written, the `rpy` path is less accurate than the eigenvalue decomposition it was meant to avoid.

Building the rotation in double precision fixes it completely. I tried a `float3_to_quatd` sibling in `utils.py`, used only here:

```
max abs error : 8.882e-16   (was 4.269e-08)
trace drift   : +0.000e+00  (was +8.412e-08)
```

Verified end-to-end on `link_box7`; `testPhysicsInertia` still passes. Identity `rpy` is an exact passthrough either way, so the no-`origin` path is unaffected.

File: urdf_usd_converter/_impl/link.py [hunk 3]
```diff
@@ -130,18 +155,10 @@ def apply_inertial(prim: Usd.Prim, link: ElementLink, data: ConversionData):
         orientation, diag_inertia = extract_inertia(inertia)
         mass_api.GetPrincipalAxesAttr().Set(orientation)
```
Comment (andrewkaufman) [inline urdf_usd_converter/_impl/link.py:161]: Minor / possible follow-up rather than something for this PR.

`physics:principalAxes` is still derived by eigendecomposing the **unrotated** tensor and then composing the `rpy` quaternion on top, while `newton:inertia` now comes from rotating the tensor directly. Two different rotation paths to the same physical quantity — they currently reconstruct tensors that differ by `1.4e-07`. Consumers taking different branches of Newton's `newton:inertia > diagonalInertia` precedence get slightly different dynamics.

Eigendecomposing `I_body` directly would make the body-frame tensor the single source of truth for both. `physics:diagonalInertia` wouldn't change at all (eigenvalues are rotation-invariant, confirmed identical either way) — only how `principalAxes` is derived.

File: tests/testPhysicsInertia.py [hunk 4]
```diff
@@ -178,5 +178,14 @@ def test_physics_inertia(self):
         self.assertAlmostEqual(mass_api.GetMassAttr().Get(), 0.8, places=6)
         self.assertTrue(Gf.IsClose(mass_api.GetDiagonalInertiaAttr().Get(), Gf.Vec3f(0.77679752, 2.1640169, 3.0591856), 1e-6))
         self.assertRotationsAlmostEqual(mass_api.GetPrincipalAxesAttr().Get(), Gf.Quatf(-0.0463736, 0.2748650, 0.9481796, 0.1524932))
-        self._assert_newton_inertia(link_box7_prim, 2.0, 1.0, 3.0, 0.5, 0.25, -0.1)
+        # `newton:inertia` is in the body/link frame (URDF inertia rotated by origin.rpy).
+        self._assert_newton_inertia(
+            link_box7_prim,
+            1.1744663749117963,
```
Comment (andrewkaufman) [inline tests/testPhysicsInertia.py]: `link_box7` is a good fixture for this — all three products of inertia non-zero and a general `rpy`, so `R I R^T` and `R^T I R` differ in every component. That's what makes it able to catch a transposed `R` at all.

Two notes:

1. These constants are the float32-rounded values, so `places=6` is currently masking the precision loss noted in `link.py`. With a double-precision rotation they'd become the true float64 tensor:

```
1.1744663602719603, 1.8959333132880136, 2.9296003264400263,
0.59298103264391533, 0.20474722009891122, -0.38820851938707363
```

2. Asserting against six hardcoded constants means that if someone flips the transform direction later, the test fails without indicating *why*. Consider reconstructing the expected tensor independently from `rpy` — the way `_assert_inertia_reconstructs` already does on the line below for the principal-axes path — so the failure message points at the convention rather than at six magic numbers.

File: docs/concept_mapping.md [hunk 5]
```diff
@@ -407,18 +407,24 @@ The inertia element contains 6 named properties, which can be used to construct
...
 
-However, in Newton USD Schemas there is `NewtonMassAPI` which provides a `newton:inertia` attribute that exactly matches the layout of the URDF inertia element as `[Ixx, Iyy, Izz, Ixy, Ixz, Iyz]`. Both representations are authored: `physics:principalAxes` and `physics:diagonalInertia` for UsdPhysics consumers, and `newton:inertia` for the symmetric inertia tensor in URDF layout.
+- `physics:principalAxes` and `physics:diagonalInertia` — computed via eigenvalue decomposition of the URDF inertia tensor, then composing `origin.rpy` into `physics:principalAxes` as described above.
+- `newton:inertia` (`NewtonMassAPI`) — the full symmetric inertia tensor as `[Ixx, Iyy, Izz, Ixy, Ixz, Iyz]` **in the body's local (link) frame**.
 
-| URDF | OpenUSD | Description |
+Because `newton:inertia` must be in the body frame, the URDF values must not be copied verbatim when `origin.rpy` is non-identity.  
+Rotate the URDF tensor into the link frame as `I_body = R * I_urdf * R^T`, then author `newton:inertia` from `I_body` as `[Ixx, Iyy, Izz, Ixy, Ixz, Iyz]`.
```
Comment (andrewkaufman) [inline docs/concept_mapping.md:416]: This is the correct form and matches what the code actually computes. Note it disagrees with the `_inertia_tensor_in_body_frame` docstring, which states `I_body = R^T * I_urdf * R` — flagged over there.

Also worth stating here that `R` is built from URDF's fixed-axis (extrinsic) XYZ convention, `R = Rz(yaw) @ Ry(pitch) @ Rx(roll)`, since that's the assumption the whole transform rests on.

**Review by andrewkaufman — COMMENTED** [review]
Reviewed the math numerically rather than by reading conventions, since the row-vector/column-vector question here is easy to get backwards.

**The core transform is correct.** `Gf` is row-vector, so `Gf.Matrix3d(quat)` is the transpose of the textbook column-convention `R`. I verified against a hand-built URDF fixed-axis `R = Rz(yaw) @ Ry(pitch) @ Rx(roll)`:

```
Gf.Matrix3d(quat) == R_col^T   -> True
code == R I R^T   (correct)    -> True
code == R^T I R   (inverse)    -> False
```

Trace is preserved (6.0), eigenvalues are unchanged (`[0.7768, 2.1640, 3.0592]` before and after, as required since a similarity transform cannot change principal moments), and the result stays symmetric. 130 tests passing locally.

One substantive issue (`Gf.Quatf` discards most of the precision this attribute exists to preserve) plus some smaller notes are inline.

File: urdf_usd_converter/_impl/link.py (hunk 1 above)
Comment (ft-lab) [reply urdf_usd_converter/_impl/link.py]: Fixed the docstring for `_inertia_tensor_in_body_frame`.

File: docs/concept_mapping.md (hunk 5 above)
Comment (ft-lab) [reply docs/concept_mapping.md:416]: I added text regarding `R = Rz(yaw) * Ry(pitch) * Rx(roll)` and the fact that `newton:inertia` should be calculated using double precision.

File: urdf_usd_converter/_impl/link.py (hunk 2 above)
Comment (ft-lab) [reply urdf_usd_converter/_impl/link.py]: I replaced the contents of `link.py` - `_inertia_tensor_in_body_frame` as follows.  

```py
    if origin:
        rotation = Gf.Matrix3d(float3_to_quatd(origin.get_with_default("rpy")))
        mat = rotation.GetTranspose() * mat * rotation
```

`float3_to_quatf` -> `float3_to_quatd`

I added `float3_to_quatd` to `utils.py`.
This is a new addition; it does not replace `float3_to_quatf`.

File: tests/testPhysicsInertia.py (hunk 4 above)
Comment (ft-lab) [reply tests/testPhysicsInertia.py]: I modified the `inertia` test for when using `origin.rpy` to match the other implementations of `test_physics_inertia`.  

```py
        # `newton:inertia` is in the body/link frame (URDF inertia rotated by origin.rpy).
        self._assert_newton_inertia(link_box7_prim, 2.0, 1.0, 3.0, 0.5, 0.25, -0.1, rpy=(0.3, -0.4, 0.5))
        self._assert_inertia_reconstructs(link_box7_prim, 2.0, 1.0, 3.0, 0.5, 0.25, -0.1, rpy=(0.3, -0.4, 0.5))
```

File: urdf_usd_converter/_impl/link.py (hunk 3 above)
Comment (ft-lab) [reply urdf_usd_converter/_impl/link.py:161]: This matter has been split off into "Derive physics:principalAxes from body-frame newton:inertia (Issue #135)."

File: urdf_usd_converter/_impl/utils.py [hunk 6]
```diff
@@ -42,6 +43,20 @@ def float3_to_quatf(rpy: tuple[float, float, float]) -> Gf.Quatf:
     return Gf.Quatf(rotation.GetQuat())
 
 
+def float3_to_quatd(rpy: tuple[float, float, float]) -> Gf.Quatd:
```
Comment (andrewkaufman) [inline urdf_usd_converter/_impl/utils.py:41]: `float3_to_quatd` duplicated the body of `float3_to_quatf`, so the URDF fixed-axis XYZ rpy convention was defined in two places and could drift.

Pushed `4729ef2` to have `float3_to_quatf` delegate, narrowing precision once at the end:

```py
def float3_to_quatf(rpy: tuple[float, float, float]) -> Gf.Quatf:
    return Gf.Quatf(float3_to_quatd(rpy))
```

Not a behavior change: bit-identical to the previous implementation across 5000 random rpy triples plus the +/-pi edge cases (max component delta `0.0`). 130/130 tests pass, ruff clean.

**Review by andrewkaufman — DISMISSED** [review]
Re-reviewed the updated head. **All four points from the previous review are addressed correctly**, and I re-verified the math numerically rather than by re-reading the diff.

Double-precision rotation confirmed on `link_box7` (`rpy=(0.3, -0.4, 0.5)`):

```
                max abs error    trace drift
quatd (now)     8.882e-16        +0.000e+00
quatf (before)  4.269e-08        +8.412e-08
```

The tensor is now authored at full `double[]` precision, the trace is exactly invariant as a similarity transform requires, eigenvalues are unchanged (`[0.7768, 2.1640, 3.0592]`), and identity `rpy` remains an exact passthrough.

Transform direction re-confirmed against a hand-built fixed-axis `R = Rz(yaw) @ Ry(pitch) @ Rx(roll)`: `code == R I R^T` -> True, `code == R^T I R` -> False.

The reworked test is now a real guard rather than six magic numbers. Two negative controls against it:

```
revert quatd -> quatf   -> FAILS (1.4640e-08 > places=12)
flip to R^T I R         -> FAILS (delta 1.2621)
```

So `places=12` catches the precision regression and the derived expectation catches a transposed `R`.

Docs and the `_inertia_tensor_in_body_frame` docstring now agree with the code, and the `physics:principalAxes` divergence is correctly split out to #135.

I pushed one small cleanup (`4729ef2`) to deduplicate the rpy conversion in `utils.py` — details inline. 130/130 tests pass, ruff clean.

LGTM.

File: urdf_usd_converter/_impl/utils.py (hunk 6 above)
Comment (andrewkaufman) [reply urdf_usd_converter/_impl/utils.py:41]: I pushed this change

**Review by andrewkaufman — APPROVED** [review]
